### ISSUE TITLE
Blockungsexport

### ISSUE THREAD

**Issue by ws46** [SVWS-NRW/SVWS-Server#327]
Ich habe versucht, über die REST-Schnittstelle eine Blockung in die DB zu exportieren. Das ist misslungen. Der Dateiname, wie verlangt, im Body.

Hier die Rückmeldung:

https://localhost/db/..../datenaustausch/gost/kurs42/import/zip
REST-Fehler: 415, Unsupported Media Type
C:\Users\Walter\AppData\Local\Temp\Kurs42.zip

Bezieht sich die Fehlermeldung auf das Format der zip-Datei oder das Format einer der in der Datei enthaltenen Textdateien?

**Comment by hmt** [comment]
Der Export bzw Import bezieht sich auf diese Schnittstelle: `debug/index.html#/Server/importKurs42Blockung`?
Der Fehler kommt mir bekannt vor, da stimmt wahrscheinlich etwas nicht mit der Übergabe. Ich jetzt nicht geschaut, wie das außerhalb von JavaScript gemacht wird, aber ich erstelle in `FormData`-Objekt und setze die Datei auf den Schlüssel `data`:
```ts
const formData = new FormData();
formData.append("data", file);
```

**Comment by FPfotenhauer** [comment]
Ich würde das Issue schließen, wenn hier keine weiteren Informationen getauscht werden.

**Comment by FPfotenhauer** [comment]
Wir verstehen das Problem nicht, da es bei uns ja funktioniert.
Warum ist der Hinweis von @hmt nicht hilfreich?

**Comment by hmt** [comment]
Der Hinweis war der, dass die zip über das Feld `data` transportiert werden muss als `multipart/formdata`.

Klappt es denn über die Debug-Schnittstelle? Oder kommt der Fehler auch darüber? `Unsupported Media Type` klingt erstmal nach falscher Übertragung, nicht falscher Inhalt. Ich kann es hier so nicht reproduzieren.

**Comment by ws46** [comment]
> Der Hinweis war der, dass die zip über das Feld `data` transportiert werden muss als `multipart/formdata`.

Das hatte ich so nicht verstanden. Über die Debug-Schnittstelle klappt es. Das hilft mir jetzt weiter.

**Comment by hmt** [comment]
wenn man im Firefox die Debug-Schnittstelle verwendet, dann kann man über Netzwerk in den Developer-Tools recht gut nachvollziehen, was passiert und welche Daten verschickt werden.

**Comment by ws46** [comment]
Der Mediatype-Fehler ist beseitigt, dafür erhalte ich jetzt die Antwort
>"id":null,"success":false,"log":["Erstelle eine temporäres Verzechnis mit dem Inhalt der Zip-Datei unter dem Namen
 \"C:\\ProgramData\\SVWS-Server/temp/svwsdb_ZSV80wuTxIVPPaerHNnQzft905B9rENQxKP78fIO.zip\"","
 [FEHLER] Unerwarteter Fehler: Cannot read the array length because \"buf\" is null"]}

Das temporäre Verzeichnist wird angelegt, bleibt aber leer. Ich hätte vermutet, dass das Verzeichnis vom SVWS-Server mit den Dateien der übergebenen Zip-Datei gefüllt wird.

**Comment by ws46** [comment]
Es wird besser, aber Fehler bleiben. Zur Verdeutlichung ein Auszug aus dem Log:
>Rest-Antwort:  400, Bad Request
{
  "id":null,
  "success":false,
  "log":["Erstelle eine temporäres Verzechnis mit dem Inhalt der Zip-Datei unter dem Namen 
  \"C:\\ProgramData\\SVWS-Server/temp/svwsdb_XVLNNfvnFFvGEtygYkShcyXER6sBg1XYkUhCDIwG.zip\"",
  "Importiere die Blockung mithilfe der extrahierten Daten:",
  "-> Lese Informationen zu der Schule ein...","  [OK]",
  "-> Bestimme die Schüler der Schule aus der Datenbank...","  [OK]",
  "-> Bestimme die Lehrkräfte der Schule aus der Datenbank...","  [OK]",
  "-> Bestimme die Fächer der Schule aus der Datenbank...","  [OK]",
  "-> Lese die Kurs 42 - Export - Textdateien ein...","  [Fehler] - Fehler beim Einlesen des Kurs42-Text-Exports: ",
  "  java.nio.file.NoSuchFileException: C:\\ProgramData\\SVWS-Server\\temp\\svwsdb_XVLNNfvnFFvGEtygYkShcyXER6sBg1XYkUhCDIwG.zip\\Blockung.txt

Die Datei existiert definitiv. Kann es sein, dass von Java durch den Backslash vor temp "\t" als Tabulator interpretiert wird?

**Comment by hmt** [comment]
Verstehe ich das richtig, wenn die Daten über die Debug-Schnittstelle verschickt werden, klappt es, wenn es mit einer eigenen Methode über die API verschickt wird, geht es nicht?
Sobald die Daten auf dem Server angekommen sind, sollte es eigentlich egal sein, wie sie verschickt wurden. Dies sieht mir nach einem Zugriffsfehler aus. Kann der Server in dem Verzeichnis die Daten lesen? Schreiben kann er ja offensichtlich, wenn die Datei vorhanden ist. Das `\t` dürfte kein Problem sein.

**Comment by ws46** [comment]
> Verstehe ich das richtig, wenn die Daten über die Debug-Schnittstelle verschickt werden, klappt es

Ja. Der Export der identischen Datei aus Kurs 42 bricht mit der o.a. Fehlermeldung ab. In den Verzeichnissen (die ja bei der Installation des Servers angelegt wurden) bestehen alle notwendigen Rechte. Ausführung des Programmes als Administrator bringt keine Änderung. Ich bin ratlos.

**Comment by ws46** [comment]
Das liefert das Ergebnislog:

![Kursexport](https://github.com/user-attachments/assets/a6a946e4-0a50-4730-94f5-32fa5a3465a5)

**Comment by hmt** [comment]
wurde der Pfad in der `svwsconfig.json` festgelegt? Mit `TempPath` kann man einen eigenen Pfad bestimmen. Das ist der Pfad, der für die zips genutzt wird. Warum die slashes und backslashes gemischt werden, erschließt sich mir nicht ganz aus dem Java-Code. Probieren sie mal einen anderen Pfad und geben sie ihn absolut ein.

Der für den Pfad verantworliche Code ist dieser hier, da fällt mir nichts dran auf: 
```java
			final String tmpDirectory = SVWSKonfiguration.get().getTempPath();
			final String tmpFilename = conn.getUser().getConfig().getDBSchema() + "_" + _random.ints(48, 123)  // from 0 to z
					.filter(i -> ((i <= 57) || (i >= 65)) && ((i <= 90) || (i >= 97)))  // filter some unicode characters
					.limit(40)
					.collect(StringBuilder::new, StringBuilder::appendCodePoint, StringBuilder::append)
					.toString() + ".zip";
			logger.logLn("Erstelle eine temporäres Verzechnis mit dem Inhalt der Zip-Datei unter dem Namen \"" + tmpDirectory + "/" + tmpFilename + "\"");
			final Path path = Paths.get(tmpDirectory).resolve(tmpFilename);
```
Dieser Pfad wird dann später mit der `Blockung.txt` aufgelöst und wirft den Fehler.

**Comment by ws46** [comment]
> wurde der Pfad in der `svwsconfig.json` festgelegt? Mit `TempPath` kann man einen eigenen Pfad bestimmen.

Ja, aber das hilft auch nicht. Ich habe den Pfad jetzt mal so festgelegt:
`c:\\Users\\Walter\\AppData\\Local\\Temp\\SVWSTest\\`

Die Debug-Schnittstelle macht daraus
"Erstelle eine temporäres Verzechnis mit dem Inhalt der Zip-Datei unter dem Namen \"c:\\Users\\Walter\\AppData\\Local\\Temp\\SVWSTest/svwsdb_T...u.zip\""
Beim Aufruf aus Kurs42 wird daraus
"Erstelle eine temporäres Verzechnis mit dem Inhalt der Zip-Datei unter dem Namen \"c:\\Users\\Walter\\AppData\\Local\\Temp\\SVWSTest\\/svwsdb_q...Y.zip\""

Im ersten Fall werden die zwei Backslashes am Ende von TempPath (korrekt) durch einen Slash ersetzt, im zweiten Fall durch  Backslash plus Slash. Übrigens wird, wenn bereits in TempPath, wie im Opriginal (`C:\\ProgramData\\SVWS-Server/temp`), ein Slash vorkommt, dieser ebenfalls durch Backslash plus Slash ersetzt. 

Ich könnte mir vorstellen, dass die fehlerhafte Behandlung der Separatoren hier
`final Path path = Paths.get(tmpDirectory).resolve(tmpFilename);`
passiert, aber das können Sie besser beurteilen.

**Comment by hmt** [comment]
lassen Sie mal die abschließenden Slashes weg am Ende von `tempPath`. Das sollte das Problem beheben.
Also: `c:\\Users\\Walter\\AppData\\Local\\Temp\\SVWSTest` (wieso eigentlich immer zwei? Ist das etwas, was man unter Windows so macht?)

**Comment by kroerig** [comment]
Nein,  normalerweise nicht. Ich kenne das aber auch nur nur, dass Java-Anwendungen Windows-Pfade immer mit Doppel-Backslash brauchen, weil ein Backslash als Escape-Zeichen interpretiert wird.

**Comment by ws46** [comment]
> lassen Sie mal die abschließenden Slashes weg am Ende von `tempPath`. Das sollte das Problem beheben.

Habe ich gemacht, ändert aber nichts.
Ich habe mal [hier](https://docs.oracle.com/javase/8/docs/api/java/nio/file/Paths.html) nachgesehen. Offenbar bauen get und resolve in Abhängigkeit vom Filesystem den Pfad zusammen. Wie das geschieht, weiß ich nicht, dafür verstehe ich zu wenig von Java.

**Comment by FPfotenhauer** [comment]
![Image](https://github.com/user-attachments/assets/fb6eb7a4-ee62-4062-b13f-05b55a70332d)

Bei mir klappt der Import mit diesen Einstellungen im Body einwandfrei.

**Comment by ws46** [comment]
> Bei mir klappt der Import mit diesen Einstellungen im Body einwandfrei.

Das habe ich auch nie bezweifelt. Dennoch ist es so, dass der Import aus Kurs 42 mit identischen Einstellungen, offenbar wegen des nicht korrekt gebildeten tempörären  Pfades, scheitert. Ich kann auch die gleiche Datei über das Debug-Inteface importieren, obwohl dabei andere Fehler auftreten. Aber das ist eine andere Sache.

**Comment by ThomasBachran** [comment]
> Kann man den Code nicht dahingehend ändern, dass die zwei letzten Zeilen vertauscht werden? Dann wüsste man, wie der ezeugte Pfad genau aussieht.
> 
> ```
> final Path path = Paths.get(tmpDirectory).resolve(tmpFilename);
> logger.logLn("Erstelle ein temporäres Verzeichnis mit dem Inhalt der Zip-Datei unter dem Namen \"" + path.toString() + "\"");
> ```

Das vertauschen ist hier nicht möglich, da die untere Zeile von der oberen abhängt. Insgesamt ist es interessant, dass sich das Verhalten zwischen Delphi und anderen Schnittstellen unterscheidet.

Interessant ist auch, dass bei dem Log zu dem Aufruf aus Delphi heraus. die forward-slashes escaped sind. Das ist laut RFC 8259 in Punkt 7 für JSON optional. Unsere Java-Implementierung macht dies z.B. per Default nicht. In der svwsconfig.json habe ich das auch nicht gesehen, so dass dieses Verhalten erstmal irritierend ist. Wenn ich das korrekt im Kopf habe, dann fügt Delphi bei forward slashes ein back-slash als escape character hinzu. Könnte das etwas damit zu tun haben?

Ansonsten können wir das Verhalten bei uns leider in mehreren Konstellationen nicht nachstellen. Allerdings wird in unserem Projekt auch kein Delphi verwendet...

**Comment by ws46** [comment]
> Interessant ist auch, dass bei dem Log zu dem Aufruf aus Delphi heraus. die forward-slashes escaped sind. Das ist laut RFC 8259 in Punkt 7 für JSON optional. Unsere Java-Implementierung macht dies z.B. per Default nicht.

Delphi macht das auch nicht und hat mit der Sache, wenn ich das richtig sehe, auch nichts zu tun, weil der temporäre Pfad vom Server und nicht von Kurs 42 angelegt wird.

**Comment by FPfotenhauer** [comment]
Was steht denn unter "TempPath" in der svwsconfig.json des SVWS-Servers?

**Comment by FPfotenhauer** [comment]
Kannst du bitte mal probieren den TempPath in der svwsconfig.json zu verändern.
Default ist "C:\\ProgramData\\SVWS-Server/temp". Danach den ServerDienst einmal neu starten.
Das wird unter Windows von Inno-Setup so gemacht.
Das ist der einzige Unterschied, den ich hier noch erkennen kann.

**Comment by FPfotenhauer** [comment]
Letzte Überlegung für heute:
Wie lautet der Dateipfad zur Originaldatei, den Kurs42 an den REST-Aufruf übergibt?

**Comment by FPfotenhauer** [comment]
```
{
  "id": null,
  "success": false,
  "log": [
    "Erstelle eine temporäres Verzechnis mit dem Inhalt der Zip-Datei unter dem Namen \"./Temp/svwsdb_lIwXeoD8uO23APFwmxREKbbAUyeVX8LO8BOdWKXe.zip\"",
    "Importiere die Blockung mithilfe der extrahierten Daten:",
    "-> Lese Informationen zu der Schule ein...",
    "  [OK]",
    "-> Bestimme die Schüler der Schule aus der Datenbank...",
    "  [OK]",
    "-> Bestimme die Lehrkräfte der Schule aus der Datenbank...",
    "  [OK]",
    "-> Bestimme die Fächer der Schule aus der Datenbank...",
    "  [OK]",
    "-> Lese die Kurs 42 - Export - Textdateien ein...",
    "  [Fehler] - Fehler beim Einlesen des Kurs42-Text-Exports: ",
    "  java.nio.file.NoSuchFileException: ./Temp/svwsdb_lIwXeoD8uO23APFwmxREKbbAUyeVX8LO8BOdWKXe.zip/Blockung.txt\n\tat java.base/sun.nio.fs.UnixException.translateToIOException(UnixException.java:92)\n
....
```

Diesen Fehler bekomme ich, wenn ich in dem ZIP einen Unterordner anlege und die txts darin sind und nicht auf der obersten Ebene im ZIP.

**Comment by ws46** [comment]
> Wie lautet der Dateipfad zur Originaldatei,

c:\Users\Walter\AppData\Local\Temp\Kurs42.zip

Die Zip-Datei enthält **keine** Unterordner. Bei Bedarf kann ich die gerne mal mischicken.

**Comment by ws46** [comment]
> Kannst du bitte mal probieren den TempPath in der svwsconfig.json zu verändern.

Habe ich alles schon gemacht. Keine Änderung.

**Comment by ws46** [comment]
> Das Vertauschen ist hier nicht möglich, da die untere Zeile von der oberen abhängt. Insgesamt ist es interessant, dass sich das Verhalten zwischen Delphi und anderen Schnittstellen unterscheidet.

Das verstehe ich nicht. Wieso beeinflusst die Bildung des Strings für die Logausgabe die Berechnung des Pfades?

>logger.logLn("Erstelle eine temporäres Verzechnis mit dem Inhalt der Zip-Datei unter dem Namen \"" + tmpDirectory + "/" + tmpFilename + "\"");
			final Path path = Paths.get(tmpDirectory).resolve(tmpFilename);

Mein Vorschlag bedeutet nur, die Logausgabe **_nach_** der endgültigen Berechnung des Pfades zu erzeugen.

**Comment by FPfotenhauer** [comment]
Es liegt nicht an den Pfadangaben.
Die ZIP-Datei wird im temp-Verzeichnis des SVWS-Servers gefunden.
Aber beim Entpacken geht etwas schief und die erste Datei die verarbeitet werden soll "Blockung.txt" in dieser ZIP wird nicht gefunden.
Deshalb hatte ich das mit dem Unterordner probiert. Es kann ja auch eine andere Ursache haben.

**Comment by kroerig** [comment]
Mal ne blöde Frage: Wie lautet der volle Pfad zur Blockung.txt?
Und: Könnte Groß-/ Kleinschreibung in den Dateinamen eine Rolle spielen?

**Comment by ws46** [comment]
> Wie lautet der volle Pfad zur Blockung.txt?

Es gibt keine Pfade in der Zip-Datei, wie man dem Directory  entnehmen kann:
28.01.2025  17:19               162 Blockgruppen.txt
28.01.2025  17:19               919 Blockplan.txt
28.01.2025  17:19             1.166 Blockung.txt
28.01.2025  17:19           119.681 Fachwahlen.txt
28.01.2025  17:19             1.220 Faecher.txt
28.01.2025  17:19               593 Jahrgaenge.txt
28.01.2025  17:19             3.345 Klassen.txt
28.01.2025  17:19             2.364 Klausuren.xml
28.01.2025  17:19            71.481 Kursbelegung.txt
28.01.2025  17:19             3.982 Kurse.txt
28.01.2025  17:19                47 Kursplaene.txt
28.01.2025  17:19            20.339 Lehrer.txt
28.01.2025  17:19               241 Raeume.txt
28.01.2025  17:19             2.061 Schienen.txt
28.01.2025  17:19            15.282 Schueler.txt
28.01.2025  17:19             4.824 SchulLogo.jpg
28.01.2025  17:19               490 ZulKurse.txt

>Und: Könnte Groß-/ Kleinschreibung in den Dateinamen eine Rolle spielen

Das muss die Server-Seite beantworten.
Ich glaube aber nicht, weil die als fehlend bemängelte Datei (Blockung.txt) in der Fehlermeldung genau so geschrieben ist wie im Verzeichnis.

Ich habe die anonymisierte (ansonsten echte) Blockungsdatei mal angehängt:
[Kurs42.zip](https://github.com/user-attachments/files/18628967/Kurs42.zip)

**Comment by kroerig** [comment]
Ich meinte den Pfad, wo der SVWS die Datei sucht. Und zwar vom C: an.

Welche Datenbank brauche ich um die Datei einlesen zu können?

**Comment by ws46** [comment]
> Ich meinte den Pfad, wo der SVWS die Datei sucht. Und zwar vom C: an.

hier: c:\Users\Walter\AppData\Local\Temp\Kurs42.zip. Das scheint ja auch kein Problem zu sein, weil der Server ja die Datei, wie FrankPfotenhauer schrieb, offenbar findet.

> Welche Datenbank brauche ich um die Datei einlesen zu können? Und ohne die Blockunsgdatei geht es ja auch nicht.

Ich kann gerne eine anonymisierte Access-DB mit zugehöriger Blockung und Programm auf meiner Homepage ablegen, wenn das weiterhilft.

**Comment by ws46** [comment]
> warum es über die SwaggerUI und mit Bruno funktioniert. 

Wer oder was ist Bruno?

> Dürfte es dann nämlich auch nicht.

Sehe ich auch so. Deshalb mein Angebot. Möglicherweise kann man das dann mal im Debugger laufen lassen.

**Comment by hmt** [comment]
> Wer oder was ist Bruno?

https://www.usebruno.com/ -> Ein Api-Client

**Comment by FPfotenhauer** [comment]
Bitte wieder öffnen, falls es noch mehr oder andere Informationen gibt.
